parser.mly

```diff
@@ -27,23 +27,60 @@ let scope_annotation (mode, xs) =
 program: optional_newlines prog=program_code EOF { prog }
 
 program_code:
-| prog=list(instruction_line)
+| l1=instruction_line prog=list(instruction_line) fs=list(afunction)
   {
     let annotations, instructions = List.split prog in
-    [("main",
-      (Array.of_list instructions,
-       Array.of_list annotations))]
+    let a1, i1 = l1 in
+    { main = {
+          name = "main";
+          formals = [];
+          body = [("anon", Array.of_list (i1::instructions))];
+          annotations = [("anon", Array.of_list (a1::annotations))]; };
+      functions = fs;
+    }
+  }
+| f1=afunction fs=list(afunction)
+  {
+    let fs = f1 :: fs in
+    let main = (try List.find (fun {name = n} -> n = "main") fs with
+        | Not_found -> (Printf.printf ("missing main function\n"); exit 1)
+      ) in
+    let rest = List.filter (fun {name = n} -> n <> "main") fs in
+    { main = main;
+      functions = rest; }
+  }
+
+formal_param:
+| CONST x=variable
+    { ParamConst x }
+| MUT x=variable
+    { ParamMut x }
+
+afunction:
+| FUNCTION name=variable LPAREN formals=separated_list(COMMA, formal_param) RPAREN NEWLINE optional_newlines prog=list(instruction_line)
+  {
+    let annotations, instructions = List.split prog in
```

Comment:
I'm not convinced by the idea of keeping annotations and instructions separate here (it means the type-checker does not prevent you from adding one without the other). One thing you could do is to have the type of the `body` field be a type parameter, and have the parser a `(annotation * instruction_stream)`-parametrized value, and have the scope checker return just an `instruction_stream`-parametrized value.